Apple/NetworkExtension/BurrowIpc.swift

```diff
@@ -0,0 +1,152 @@
+            guard let (lines, size) = lines(from: buffer) else {
+                return 0
+            }
+            result = lines
+            return size
+        }
+        for line in result {
+            framer.deliverInput(data: line, message: .init(instance: framer), isComplete: true)
+        }
+        return 0
+    }
+    func handleOutput(
+        framer: NWProtocolFramer.Instance,
+        message: NWProtocolFramer.Message,
+        messageLength: Int,
+        isComplete: Bool
+    ) {
+        do {
+            try framer.writeOutputNoCopy(length: messageLength)
+        } catch {
+        }
+    }
+}
+
+extension NWConnection {
+    func receiveMessage() async throws -> (Data?, NWConnection.ContentContext?, Bool) {
+        try await withUnsafeThrowingContinuation { continuation in
+            receiveMessage { completeContent, contentContext, isComplete, error in
+                if let error = error {
+                    continuation.resume(throwing: error)
+                }
+                continuation.resume(returning: (completeContent, contentContext, isComplete))
+            }
+        }
+    }
+}
+
+final class BurrowIpc {
+    let connection: NWConnection
+    private var generator = SystemRandomNumberGenerator()
+    private var continuations: [UInt: UnsafeContinuation<Data, Error>] = [:]
+    private var logger: Logger
+    init(logger: Logger) {
+        let params = NWParameters.tcp
+        params.defaultProtocolStack
+            .applicationProtocols
+            .insert(NWProtocolFramer.Options(definition: LineProtocol.definition), at: 0)
+        let connection = NWConnection(to: .unix(path: "burrow.sock"), using: params)
+        connection.start(queue: .global())
+        self.connection = connection
+        self.logger = logger
+    }
+    func send<T: Request, U: Decodable>(_ request: T) async throws -> U {
+        let data: Data = try await withUnsafeThrowingContinuation { continuation in
+            let id: UInt = generator.next(upperBound: UInt.max)
+            continuations[id] = continuation
+            var copy = request
+            copy.id = id
+            do {
+                var data = try JSONEncoder().encode(request)
+                data.append(contentsOf: [10])
+                let completion: NWConnection.SendCompletion = .contentProcessed { error in
+                    guard let error = error else { return }
+                    continuation.resume(throwing: error)
+                }
+                connection.send(content: data, completion: completion)
+            } catch {
+                continuation.resume(throwing: error)
+                return
+            }
+        }
+        return try JSONDecoder().decode(Response<U>.self, from: data).result
+    }
+    func send_raw(_ request: Data) async throws -> Data {
+        try await withCheckedThrowingContinuation { continuation in
```

Comment:
We should move the continuation to an extension so that we can keep `BurrowIpc` more readable

Reply:
Agreed!
As I understand it, should we refactor the continuations inside `BurrowIpc` as functions in `extension NWConnection`? Moving the continuations into an extension of `BurrowIpc` itself might require moving all the functions that depend on it into the extension.